pcapng/structs.py

```diff
@@ -470,42 +698,92 @@ class Options(Mapping):
         Required in order to load numeric fields.
     """
 
+    __slots__ = [
+        "schema",
+        "_field_names",
+        "data",
+        "endianness",
+    ]
+
     def __init__(self, schema, data, endianness):
-        self.schema = {}  # Schema of option fields: {<code>: {..def..}}
+        self.schema = {}  # Schema of option fields: {<code>: Option(...)}
         self._field_names = {}  # Map names to codes
-        self.raw_data = {}  # List of (code, value) tuples
+        self.data = defaultdict(list)  # option data, with numeric option IDs as keys
         self.endianness = endianness  # one of '<>!='
 
         # This is the default schema, common to all objects
-        self._update_schema([(0, "opt_endofopt"), (1, "opt_comment", TYPE_STRING)])
-        self._update_schema(schema)
+        for item in [
+            Option(0, "opt_endofopt"),
+            Option(1, "opt_comment", TYPE_STRING, multiple=True),
+            # The spec calls all these next options ``opt_custom`` --
+            # I've renamed them here so they can be told apart
+            Option(2988, "custom_str_safe", TYPE_OPT_CUSTOM_STR, multiple=True),
+            Option(2989, "custom_bytes_safe", TYPE_OPT_CUSTOM_BYTES, multiple=True),
+            Option(19372, "custom_str", TYPE_OPT_CUSTOM_STR, multiple=True),
+            Option(19373, "custom_bytes", TYPE_OPT_CUSTOM_BYTES, multiple=True),
+        ] + list(schema):
+            if not isinstance(item, Option):
+                raise TypeError("expected option, got '{}'".format(item))
+            self.schema[item.code] = item
+        self._field_names = {x.name: x.code for x in self.schema.values()}
 
         # Update raw data with current values
         self._update_data(data)
 
     # -------------------- Nice interface :) --------------------
 
+    def __eq__(self, other):
+        return self.data == other.data
+
     def __getitem__(self, name):
-        return self._get_converted(name)
+        code = self._resolve_name(name)
+        if code not in self.data:
```

Comment:
Using a `defaultdict` as you requested did result in removing a few checks. But I had to add this one instead in order to keep `tests/test_structs.py` happy. Besides raising `IndexError`s instead of `KeyError`s (since the empty list has no index #0), testing for nonexistent keys in a defaultdict causes them to be created, breaking the length check later in the test suite.